### PR TITLE
User model

### PR DESCRIPTION
**Pull request by Emiliocaamano**
## Summary

Implementation of User Model, migrations and services 

## Known Issues

None

## Trello Card

(https://trello.com/c/XZZPCAvf)

### CODE REVIEW

**Review by javerBenavidez — CHANGES_REQUESTED** [review]
You created this PR just for User Model Card? or for SingUp too?
I think that is for two cards, right?
So, if you see [this card](https://trello.com/c/1D5ybc3L/35-sign-up) you need to do some more tasks, see the checklist please to complete de PR

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  try {
+    const userCreated = await userService.createUser({
+      firstName: req.body.first_name,
+      lastName: req.body.last_name,
+      email: req.body.email,
+      password: req.body.password
+    });
+    return res.status(201).send(userCreated);
+  } catch (error) {
+    return res.status(500).send(error);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: Remember how to manage the error. 
See jokes controller.

File: app/serializers/varParse.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  try {
+    const userParsed = {
+      first_name: userParams.firstName,
+      last_name: userParams.lastName,
+      email: userParams.email,
+      password: userParams.password
+    };
+    return userParsed;
+  } catch (error) {
+    throw defaultError(errorMsg.parseError);
```
Comment (javerBenavidez) [inline app/serializers/varParse.js]: for these tasks, it is not necessary to do a `try catch` block.

File: app/services/users.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const { databaseError } = require('../errors');
+const User = require('../models').users;
+const errorMsg = require('../constants/errorMessages');
+const varParse = require('../serializers/varParse');
+
+exports.createUser = userParams => {
+  try {
+    const parsedParams = varParse.nameParse(userParams);
```
Comment (javerBenavidez) [inline app/services/users.js]: It is better to pass the userParams already parsed, so you can parse the params in the controller.

File: app/services/users.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    return User.create(parsedParams);
+  } catch (error) {
+    throw databaseError(errorMsg.userError);
```
Comment (javerBenavidez) [inline app/services/users.js:9]: Good!

File: migrations/migrations/20220906155837-user-table.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        allowNull: false,
+        validate: {
+          isAlpha: {
+            msg: 'El nombre solo puede contener letras'
+          },
+          len: {
+            args: [2, 255],
+            msg: 'El nombre tiene que tener como minimo dos caracteres'
+          }
+        }
```
Comment (javerBenavidez) [inline migrations/migrations/20220906155837-user-table.js]: I think it is not necessary to put those validations in the migration files, just put the type, allowNull, if the variables are uniques for example.

File: app/controllers/users.js [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
 
 exports.signUp = async (req, res) => {
   try {
-    const userCreated = await userService.createUser({
+    const userCreated = await varParse.nameParse({
       firstName: req.body.first_name,
       lastName: req.body.last_name,
       email: req.body.email,
       password: req.body.password
     });
```
Comment (javerBenavidez) [inline app/controllers/users.js]: - This function is not an asynchronous operation (remove async-await).
- Just send to the function `req.body`.

File: app/controllers/users.js [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
+    userService.createUser(userCreated);
     return res.status(201).send(userCreated);
   } catch (error) {
-    return res.status(500).send(error);
+    throw databaseError(errorMsg.userError);
   }
```
Comment (javerBenavidez) [inline app/controllers/users.js]: Remember hot we managed the error in the controller file.
See the jokes controller in catch block.

File: app/controllers/users.js [hunk 8]
```diff
@@ -1,15 +1,19 @@
...
   try {
-    const userCreated = await userService.createUser({
+    const userCreated = await varParse.nameParse({
       firstName: req.body.first_name,
       lastName: req.body.last_name,
       email: req.body.email,
       password: req.body.password
     });
+    userService.createUser(userCreated);
     return res.status(201).send(userCreated);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: `userService.createUser` is an asynchronous operation.

File: app/serializers/varParse.js [hunk 9]
```diff
@@ -1,16 +1,9 @@
...
-  } catch (error) {
-    throw defaultError(errorMsg.parseError);
-  }
+  const userParsed = {
+    first_name: userParams.firstName,
+    last_name: userParams.lastName,
+    email: userParams.email,
+    password: userParams.password
+  };
+  return userParsed;
```
Comment (javerBenavidez) [inline app/serializers/varParse.js]:
```suggestion
  return = {
    first_name: userParams.firstName,
    last_name: userParams.lastName,
    email: userParams.email,
    password: userParams.password
  };
```

File: app/services/users.js [hunk 10]
```diff
@@ -1,12 +1,10 @@
 const { databaseError } = require('../errors');
 const User = require('../models').users;
 const errorMsg = require('../constants/errorMessages');
-const varParse = require('../serializers/varParse');
 
 exports.createUser = userParams => {
   try {
-    const parsedParams = varParse.nameParse(userParams);
-    return User.create(parsedParams);
+    return User.create(userParams);
```
Comment (javerBenavidez) [inline app/services/users.js]: Remember to use async-await

File: app/services/users.js [hunk 11]
```diff
@@ -2,9 +2,10 @@ const { databaseError } = require('../errors');
 const User = require('../models').users;
 const errorMsg = require('../constants/errorMessages');
 
-exports.createUser = userParams => {
+exports.createUser = async userParams => {
   try {
-    return User.create(userParams);
+    await User.create(userParams);
+    return userParams;
```
Comment (javerBenavidez) [inline app/services/users.js]: Why are you returning userParams?
`return await User.create(userParams);`

File: app/controllers/users.js [hunk 12]
```diff
@@ -1,19 +1,17 @@
...
-exports.signUp = async (req, res) => {
+exports.signUp = async (req, res, next) => {
   try {
-    const userCreated = await varParse.nameParse({
+    const userCreated = varParse.nameParse({
       firstName: req.body.first_name,
       lastName: req.body.last_name,
       email: req.body.email,
       password: req.body.password
     });
```
Comment (javerBenavidez) [inline app/controllers/users.js]: just pass to the `nameParse` all the `req.body`

Comment (Emiliocaamano) [reply app/controllers/users.js]: This should be like this? 
`firstName: req.body`

File: app/controllers/users.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+const varParse = require('../serializers/varParse');
+
+exports.signUp = async (req, res, next) => {
+  try {
+    const userCreated = varParse.nameParse({
+      firstName: req.body.first_name,
+      lastName: req.body.last_name,
+      email: req.body.email,
+      password: req.body.password
+    });
```
Comment (javerBenavidez) [inline app/controllers/users.js]:
```suggestion
    const userObject = req.body;
    const userCreated = varParse.nameParse(userObject);
```

**Review by javerBenavidez — APPROVED** [review]

**Pull request merged**